summary/summary.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Maintains the summary of aggregated minute, hour, and day stats.
+// For a container running for more than a day, amount of tracked data can go up to
+// 40 KB when cpu and memory are tracked. We'll start by enabling collection for the
+// node, followed by docker, and then all containers as we understand the usage pattern
+// better
+// TODO(rjnagal): Optimize the size if we start running it for every container.
+package summary
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/cadvisor/info"
+)
+
+// Usage fields we track for generating percentiles.
+type secondSample struct {
+	Timestamp time.Time // time when the sample was recorded.
+	Cpu       uint64    // cpu usage
+	Memory    uint64    // memory usage
+}
+
+type availableResources struct {
+	Cpu    bool
+	Memory bool
+}
+
+type StatsSummary struct {
+	// Resources being tracked for this container.
+	available availableResources
+	// list of second samples. The list is cleared when a new minute samples is generated.
+	secondSamples []*secondSample
+	// minute percentiles. We track 24 * 60 maximum samples.
+	minuteSamples *SamplesBuffer
+	// latest derived minute, hour, and day stats. Updated every minute.
+	derivedStats info.DerivedStats // Guarded by dataLock.
+	dataLock     sync.RWMutex
+}
+
+// Adds a new seconds sample.
+// If enough seconds samples are collected, a minute sample is generated and derived
+// stats are updated.
+func (s *StatsSummary) AddSample(stat info.ContainerStats) error {
+	sample := secondSample{}
+	if s.available.Cpu {
+		sample.Cpu = stat.Cpu.Usage.Total
+	}
+	if s.available.Memory {
+		sample.Memory = stat.Memory.WorkingSet
+	}
+	s.secondSamples = append(s.secondSamples, &sample)
+	// TODO(jnagal): Use 'available' to avoid unnecessary computation.
```

Comment:
yes, this is a TODO. To support that, we save timestamps for each second. Instead of checking the number of samples, we need to look at timestamp. I moved it out to later because I wasn't sure how we were going to handle hour and day yet.